Fixes issue #24849 make relevant methods synchronized in LocalTxConnectionEventListener

escay opened the pull request:
Make relevant methods synchronized in LocalTxConnectionEventListener and protect associatedHandles from external clear calls.

I first tried to move most of the ConnectorXAResource.resetAssociation method inside a synchronous method in the LocalTxConnectionEventListener, but that also meant two other methods from ConnectorXAResource needed to be refactored and moved into LocalTxConnectionEventListener. So I undid that and chose for cloning the map and returning a clone to make the method getAssociatedHandles thread safe. This is a different solution as used in the Payara fix I mention in the issue. But I think it is better to leave the association and disassociating of ManagedConnection logic inside the ConnectorXAResource code.

I did not add multi threading unit tests, since only the basic synchronized(this) approach is used to protect the associatedHandles map. No need to test java. I did add some basic unit test logic for the add and remove logic used by ConnectorXAResource.

escay committed:
Fixes issue #24849 make relevant methods synchronized in LocalTxConnectionEventListener and protect associatedHandles from external clear calls.

Fixes issue #24849 make relevant methods synchronized in LocalTxConnectionEventListener and protect associatedHandles from external clear calls.

escay commented on appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/ConnectorXAResource.java:
```diff
@@ -120,7 +124,9 @@ public void end(Xid xid, int flags) throws XAException {
                 if (handle != null) { // not needed, just to be sure.
                     ManagedConnection associatedConnection = (ManagedConnection) handle.getResource();
                     associatedConnection.associateConnection(userHandle);
-                    _logger.log(FINE, "connection_sharing_reset_association", userHandle);
```
You mean I can remove the
if (_logger.isLoggable(FINE)) {

in both cases. I added it here, to be in sync / reflect the resetAssociation() original code which did use the if.
I will remove both ifs.

escay committed:
Fixes issue #24849 make relevant methods synchronized in
LocalTxConnectionEventListener and protect associatedHandles from
external clear calls. Process review comment and fix typo.

avpinchuk commented on appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/listener/LocalTxConnectionEventListener.java:
```diff
@@ -32,37 +32,51 @@
+    public synchronized void connectionClosed(ConnectionEvent evt) {
         Object connectionHandle = evt.getConnectionHandle();
         ResourceHandle handle = resource;
         if (associatedHandles.containsKey(connectionHandle)) {
```
Here we do two map lookups. Maybe it's better to use

```java
ResourceHande handle = associatedHandes.getOrDefault(connectionHandle, resource);
```

escay replied:
Thank you, I was not aware of this getOrDefault option on Maps.
Documentation of getOrDefault defines: "The default implementation makes no guarantees about synchronization or atomicity properties of this method." so I think the synchronization on the method is still needed.
I can make an update later today.

avpinchuk replied:
You're right, synchronization is needed. We simply eliminate unnecessary map access.

avpinchuk commented on appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/listener/LocalTxConnectionEventListener.java:
```diff
@@ -72,12 +86,15 @@ public void connectionErrorOccurred(ConnectionEvent evt) {
+    public synchronized void badConnectionClosed(ConnectionEvent evt) {
         Object connectionHandle = evt.getConnectionHandle();
         ResourceHandle handle = resource;
         if (associatedHandles.containsKey(connectionHandle)) {
```
Here too

avpinchuk replied:
Also, because `synchronized` is only used to guard access to the associated handles map, maybe it's better to use `ReentrantReadWriteLock` instead?

avpinchuk commented:
Also, because `synchronized` is only used to guard access to the associated handles map, maybe it's better to use `ReentrantReadWriteLock` instead?

escay commented:
> Also, because `synchronized` is only used to guard access to the associated handles map, maybe it's better to use `ReentrantReadWriteLock` instead?

I cannot oversee the impact of using ReentrantReadWriteLock on only the associatedHandles. Some code like the
 `poolManager.resourceClosed(handle);` calls could also be affected by the synchronized approach. I have no easy way to test this, I know the synchronized version works for me for the last years in Payara: https://github.com/payara/Payara/commit/c09e8aafd53221a918230db5d2c9c1098c5f2a76 to avoid the exception, so I tried to stay near that solution.
 
To know for sure I think I need to have found the real cause of issue #24805 and probably also of issue  #23483 which I think could be related and run my test set many times. 
I am also looking at: 

https://github.com/eclipse-ee4j/glassfish/blob/3e5203c56e63abf68cece6ddc76cf9ffd91c3d69/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/resource/pool/ConnectionPool.java#L679C1-L682C66

where I suspect resourceHandles can get mixed up, and could perhaps cause this issue as side effect.

escay committed:
Fixes issue #24849 make relevant methods synchronized in
LocalTxConnectionEventListener and protect associatedHandles from
external clear calls. Process review comment: use getOrDefault.
